### PR TITLE
ROI support for nvjpeg2k decoder

### PR DESCRIPTION
**Pull request by staniewzki**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Adding ROI support for nvjpeg2k decoder, by using the library's decode parameters.
It seems that the `nvjpeg2kDecodeImage` function is bugged, so decoding is done via `nvjpeg2kDecodeTile`, tile by tile.

### Affected modules and functionalities:
N/A

### Key points relevant for the review:
Decoding tile by tile in `NvJpeg2000DecoderInstance::DecodeJpeg2000`
Reviewing by commits might be easier.

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 1]
```diff
@@ -23,6 +23,19 @@
 namespace dali {
 namespace imgcodec {
 
+namespace {
+bool handle_decode_ret_status(nvjpeg2kStatus_t status, ImageSource *in) {
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:27]: nitpick: I'd shorted the name a bit: `handle_status` or so

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 2]
```diff
@@ -59,88 +72,149 @@ NvJpeg2000DecoderInstance::~NvJpeg2000DecoderInstance() {
...
+
+  if (!ctx.roi) {
+    auto output_image = PrepareOutputArea(out, pixel_data, pitch_in_bytes, 0, 0, ctx);
+    auto ret = nvjpeg2kDecode(nvjpeg2k_handle_, ctx.nvjpeg2k_decode_state,
+                              ctx.nvjpeg2k_stream, &output_image, ctx.cuda_stream);
+    return handle_decode_ret_status(ret, in);
   } else {
-    CUDA_CALL_EX(ret, in->SourceInfo());
-    DALI_FAIL("Unreachable");  // silence a warning
+    // Decode tile by tile: nvjpeg2kDecodeImage seems to be bugged
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:142]:
```suggestion
    // Decode tile by tile: nvjpeg2kDecodeImage currently doesn't work with tiled jpeg2000
```

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc (hunk 2 above)
Comment (staniewzki) [reply dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:142]: That is not the case, I couldn't get it to work with images including just one tile.

**Review by awolant — APPROVED** [review]

**Comment by staniewzki** [comment]
!build

**Pull request merged**